solidity/contracts/converter/types/liquidity-pool-v3/LiquidityPoolV3Converter.sol

```diff
@@ -0,0 +1,1356 @@
+      * @param _amount      amount of tokens received from the user
+      *
+      * @return expected target amount
+      * @return expected fee
+    */
+    function targetAmountAndFee(IERC20Token _sourceToken, IERC20Token _targetToken, uint256 _amount)
+        public
+        view
+        override
+        active
+        returns (uint256, uint256)
+    {
+        uint256 sourceId = reserveIds[_sourceToken];
+        uint256 targetId = reserveIds[_targetToken];
+
+        (uint256 sourceBalance, uint256 targetBalance) = getReserveBalances(sourceId, targetId);
+        uint256 amount = crossReserveTargetAmount(sourceBalance, targetBalance, _amount);
+
+        // return the amount minus the conversion fee and the conversion fee
+        uint256 fee = calculateFee(amount);
+        return (amount - fee, fee);
+    }
+
+    /**
+      * @dev converts a specific amount of source tokens to target tokens
+      *
+      * @param _sourceToken source ERC20 token
+      * @param _targetToken target ERC20 token
+      * @param _amount      amount of tokens to convert (in units of the source token)
+      * @param _trader      address of the caller who executed the conversion
+      * @param _beneficiary wallet to receive the conversion result
+      *
+      * @return amount of tokens received (in units of the target token)
+    */
+    function doConvert(IERC20Token _sourceToken, IERC20Token _targetToken, uint256 _amount, address _trader, address payable _beneficiary)
+        internal
+        returns (uint256)
+    {
+        // update the recent average rate
+        updateRecentAverageRateIfNeeded();
+
+        uint256 sourceId = reserveIds[_sourceToken];
+        uint256 targetId = reserveIds[_targetToken];
+
+        (uint256 sourceBalance, uint256 targetBalance) = getReserveBalances(sourceId, targetId);
+        uint256 targetAmount = crossReserveTargetAmount(sourceBalance, targetBalance, _amount);
+
+        // get the target amount minus the conversion fee and the conversion fee
+        uint256 fee = calculateFee(targetAmount);
+        uint256 amount = targetAmount - fee;
+
+        // ensure that the trade gives something in return
+        require(amount != 0, "ERR_ZERO_TARGET_AMOUNT");
+
+        // ensure that the trade won't deplete the reserve balance
+        assert(amount < targetBalance);
+
+        // ensure that the input amount was already deposited
+        uint256 actualSourceBalance;
+        if (_sourceToken == ETH_RESERVE_ADDRESS) {
+            actualSourceBalance = address(this).balance;
+            require(msg.value == _amount, "ERR_ETH_AMOUNT_MISMATCH");
+        }
+        else {
+            actualSourceBalance = _sourceToken.balanceOf(address(this));
+            require(msg.value == 0 && actualSourceBalance.sub(sourceBalance) >= _amount, "ERR_INVALID_AMOUNT");
+        }
+
+        // sync the reserve balances
+        setReserveBalance(sourceId, actualSourceBalance);
```

Comment:
these two can be updated using a single `setReserveBalance(balance0, balance1)` call - will save an additional 2 getReserveBalance calls internally

Reply:
Fixed